content/learning-paths/cross-platform/simd-info-demo/simdinfo-example2.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Intrinsics without Equivalents
+weight: 7
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+### Handling intrinsics without direct equivalents
+
+During the porting process, you will observe that certain instructions translate seamlessly. However, there are cases where direct equivalents for some intrinsics may not be readily available across architectures. For example, the [**`_mm_madd_epi16`**](https://simd.info/c_intrinsic/_mm_madd_epi16/) intrinsic from **SSE2**, which performs multiplication of 16-bit signed integer elements in a vector and then does a pairwise addition of adjacent elements increasing the element width, does not have a direct counterpart in **NEON**. However it can be emulated using another intrinsic. Similarly its 256 and 512-bit counterparts, [**`_mm256_madd_epi16`**](https://simd.info/c_intrinsic/_mm256_madd_epi16/) and [**`_mm512_madd_epi16`**](https://simd.info/c_intrinsic/_mm512_madd_epi16/) can be emulated by a sequence of instructions, but here you will see the 128-bit variant.
+
+You may already know the equivalent operations for this particular intrinsic, but let's assume you don't. In this usecase, reading the **`_mm_madd_epi16`** on the **SIMD.info** might indicate that a key characteristic of the instruction involved is the *widening* of the result elements, from 16-bit to 32-bit signed integers. Unfortunately, that is not the case, as this particular instruction does not actually increase the size of the element holding the result values. You will see how that effects the result in the example.
+
+Consider the following code for **SSE2**. Create a new file for the code named `_mm_madd_epi16_test.c` with the contents shown below:
```

Comment:
It's worth noting here, or in the general section that especially whenever there isn't a 1-1 mapping for intrinsics that you can often get better performance on Arm platforms by reorganizing the data layout to match a native Arm instruction.  As an example you often have code doing unsigned widening multiplies using `{xxx,0001,yyy,0001}` i.e. only multiplying the even bits and just zero extending the odd elements.  In this case you should permute the even/odd elements out and do the zero extend and multiple separately.  To avoid the expensive multiply.  Just a quick example but you get the idea.

Reply:
Thank you for the comment once again. Our primary focus in this work was to optimize the existing algorithm using SIMD intrinsics directly, without altering the algorithm or changing data patterns. While reordering data to align with native ARM instructions can indeed improve performance in some cases, our scope here was limited to optimizing within the constraints of the current data layout and algorithm. We mentioned this idea in the conclusion, also pointing to another LP about vectorization-friendly data layout.